### DISCUSSION TITLE
ValueError: <class 'list'> has no matching SQLAlchemy type

### DISCUSSION THREAD

**Discussion by fabiobeoni** [fastapi/sqlmodel#1248] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# hero.py (works)
class Hero(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    ....

    team_id: int | None = Field(default=None, foreign_key="team.id")
    team: Team | None = Relationship(back_populates="heroes")


# team.py (works)
class Team(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    .....

    heroes: list["Hero"] = Relationship(back_populates="team")


# todo.py (does not work)
class Todo(SQLModel, table=True):

    id: UUID | None = Field(
        default=uuid4(),
        primary_key=True,
    )

    ......

    category_id: UUID | None = Field(default=None, foreign_key="category.id")
    category: Category | None  = Relationship(back_populates="todos")


# category.py  (does not work => ValueError: <class 'list'>)
class Category(SQLModel, table=True):
    id: UUID | None = Field(default=uuid4(), primary_key=True)
    .....

    todos: list["Todo"] = Relationship(back_populates="category")
```

### Description

Given these classes and 1-to-N relationships, I get compile time error: `ValueError: <class 'list'> has no matching SQLAlchemy type` referring to the `category.py` field `todos: list["Todo"]`, while the relation hero/team does work. 

I cannot see any relevant difference between `todo/category` implementation and `hero/team` implementation to justify the error:

```
import todo.models.category
  File ".../todo/models/category.py", line 9, in <module>
    class Category(SQLModel, table=True):
  File ".../lib/python3.12/site-packages/sqlmodel/main.py", line 559, in __new__
    col = get_column_from_field(v)
          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "../lib/python3.12/site-packages/sqlmodel/main.py", line 708, in get_column_from_field
    sa_type = get_sqlalchemy_type(field)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "../lib/python3.12/site-packages/sqlmodel/main.py", line 697, in get_sqlalchemy_type
    raise ValueError(f"{type_} has no matching SQLAlchemy type")
ValueError: <class 'list'> has no matching SQLAlchemy type
``` 

### Operating System

Linux

### Operating System Details

fastapi 0.115.6 FastAPI framework, high performance, easy to learn, fast to code, ready for production
├── pydantic >=1.7.4,<1.8 || >1.8,<1.8.1 || >1.8.1,<2.0.0 || >2.0.0,<2.0.1 || >2.0.1,<2.1.0 || >2.1.0,<3.0.0
│   ├── annotated-types >=0.6.0 
│   ├── pydantic-core 2.27.2 
│   │   └── typing-extensions >=4.6.0,<4.7.0 || >4.7.0 
│   └── typing-extensions >=4.12.2 (circular dependency aborted here)
├── starlette >=0.40.0,<0.42.0
│   └── anyio >=3.4.0,<5 
│       ├── idna >=2.8 
│       ├── sniffio >=1.1 
│       └── typing-extensions >=4.5 
└── typing-extensions >=4.8.0
python-dotenv 1.0.1 Read key-value pairs from a .env file and set them as environment variables
sqlmodel 0.0.22 SQLModel, SQL databases in Python, designed for simplicity, compatibility, and robustness.
├── pydantic >=1.10.13,<3.0.0
│   ├── annotated-types >=0.6.0 
│   ├── pydantic-core 2.27.2 
│   │   └── typing-extensions >=4.6.0,<4.7.0 || >4.7.0 
│   └── typing-extensions >=4.12.2 (circular dependency aborted here)
└── sqlalchemy >=2.0.14,<2.1.0
    ├── greenlet !=0.4.17 
    └── typing-extensions >=4.6.0 

### SQLModel Version

0.0.22

### Python Version

3.12

**Comment by YuriiMotov** [comment]
Can't reproduce. Works well:

<details>

```py
from typing import Optional
import uuid
from sqlmodel import SQLModel, Field, Relationship, create_engine


class Todo(SQLModel, table=True):

    id: uuid.UUID | None = Field(
        default=uuid.uuid4(),
        primary_key=True,
    )

    category_id: uuid.UUID | None = Field(default=None, foreign_key="category.id")
    category: Optional["Category"]  = Relationship(back_populates="todos")


# category.py  (does not work => ValueError: <class 'list'>)
class Category(SQLModel, table=True):
    id: uuid.UUID | None = Field(default=uuid.uuid4(), primary_key=True)

    todos: list["Todo"] = Relationship(back_populates="category")


engine = create_engine("sqlite:///")


SQLModel.metadata.create_all(engine)

```

</details>

**Comment by alvaroalonsocano** [comment]
I have the same issue when executing unit tests and using "from __future__ import annotations" and Type_checking. Is there any good tutorial or link to follow on how to implement relationships correctly? 

I´ll paste a bit of my code so you can understand my issue:

User.py

```
from __future__ import annotations
from typing import TYPE_CHECKING

from sqlmodel import Field, Relationship, SQLModel

if TYPE_CHECKING:
    from app.models.order import Order
    from app.models.product import Product
    
class User(SQLModel, table=True):
    __tablename__ = "users"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    products: list["Product"] = Relationship(
        back_populates="owner",
    )
    orders: list["Order"] = Relationship(
        back_populates="owner",
    )
    orders_buyer: list["Order"] = Relationship(
        back_populates="buyer",
    )
   ```
   
   Order.py
   
   ```
from __future__ import annotations

from sqlalchemy import CheckConstraint
from sqlmodel import Field, Float, Relationship, SQLModel

if TYPE_CHECKING:
    from app.models.order_details import OrderDetails
    from app.models.user import User

class Order(SQLModel, table=True):
    __tablename__ = "orders"
    __table_args__ = (CheckConstraint("total_selling_price >= 0"),)

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    owner_id: uuid.UUID = Field(foreign_key="users.id", index=True)
    buyer_id: uuid.UUID = Field(foreign_key="users.id", index=True)

    owner: "User" | None = Relationship(
        back_populates="orders",
    )
    buyer: "User" | None = Relationship(
        back_populates="orders_buyer",
    )

   ```
   
   The exact error i get is:
   "sqlalchemy.exc.InvalidRequestError: One or more mappers failed to initialize - can't proceed with initialization of other mappers. Triggering mapper:.."

**Reply by YuriiMotov** [reply]
Please, fix the formatting of your code blocks. Use triple backtics for multiline blocks instead of single

**Reply by alvaroalonsocano** [reply]
> Please, fix the formatting of your code blocks. Use triple backtics for multiline blocks instead of single

Done!! Thanks in advance :)

**Reply by YuriiMotov** [reply]
I think this section of docs is what you need: https://sqlmodel.tiangolo.com/tutorial/code-structure/

Also, look at this issue: https://github.com/fastapi/sqlmodel/issues/196
